lib/IRGen/GenMeta.cpp

```diff
@@ -759,78 +748,136 @@ namespace {
-        auto call = IGF.Builder.CreateCall(IGF.IGM.getGetFunctionMetadata1Fn(),
-                                           {flags, arg0, resultMetadata});
+      case 1:
+      case 2:
+      case 3: {
+        llvm::SmallVector<llvm::Value *, 8> arguments;
+        auto *metadataFn = constructSimpleCall(arguments);
+        auto *call = IGF.Builder.CreateCall(metadataFn, arguments);
         call->setDoesNotThrow();
         return setLocal(CanType(type), call);
-        }
+      }
 
-        case 2: {
-          auto arg0 = extractAndMarkInOut(params[0]);
-          auto arg1 = extractAndMarkInOut(params[1]);
-          auto call = IGF.Builder.CreateCall(
-                                            IGF.IGM.getGetFunctionMetadata2Fn(),
-                                            {flags, arg0, arg1, resultMetadata});
-          call->setDoesNotThrow();
-          return setLocal(CanType(type), call);
-        }
+      default:
+        auto *const Int32Ptr = IGF.IGM.Int32Ty->getPointerTo();
 
-        case 3: {
-          auto arg0 = extractAndMarkInOut(params[0]);
-          auto arg1 = extractAndMarkInOut(params[1]);
-          auto arg2 = extractAndMarkInOut(params[2]);
-          auto call = IGF.Builder.CreateCall(
-                                            IGF.IGM.getGetFunctionMetadata3Fn(),
-                                            {flags, arg0, arg1, arg2,
-                                             resultMetadata});
-          call->setDoesNotThrow();
-          return setLocal(CanType(type), call);
-        }
+        llvm::SmallVector<llvm::Value *, 8> arguments;
+        arguments.push_back(flags);
 
-        default:
-          auto arrayTy = llvm::ArrayType::get(IGF.IGM.Int8PtrTy, numParams + 2);
-          Address buffer = IGF.createAlloca(arrayTy,
-                                            IGF.IGM.getPointerAlignment(),
-                                            "function-arguments");
-          IGF.Builder.CreateLifetimeStart(buffer,
+        Address parameters;
+        if (!params.empty()) {
+          parameters = getArrayFor(IGF.IGM.TypeMetadataPtrTy, numParams,
+                                   "function-parameters");
+
+          IGF.Builder.CreateLifetimeStart(parameters,
                                           IGF.IGM.getPointerSize() * numParams);
-          Address pointerToFirstArg = IGF.Builder.CreateStructGEP(buffer, 0,
-                                                                   Size(0));
-          Address flagsPtr = IGF.Builder.CreateBitCast(pointerToFirstArg,
-                                               IGF.IGM.SizeTy->getPointerTo());
-          IGF.Builder.CreateStore(flags, flagsPtr);
-
-          for (auto i : indices(params)) {
-            auto argMetadata = extractAndMarkInOut(params[i]);
-            Address argPtr = IGF.Builder.CreateStructGEP(buffer, i + 1,
+
+          ConstantInitBuilder paramFlags(IGF.IGM);
+          auto flagsArr = paramFlags.beginArray();
```

Comment:
@rjmccall One last question I have is if we should keep flags as constant array of stack allocated one? After digging into all this more it seems like we might better off with stack allocated array flags the same way as for parameters because we don't keep them around that long anyway, WDYT?

Reply:
It is almost certainly less overall binary size to use a const global array than to create one fresh on the stack.

Reply:
Ok, sounds good, I'll leave it as is then! Thanks again!